brave/brave-ios
# Fix #7765, Fix #7618: Portfolio Group By filter & Empty State

## Summary of Changes
- Implements the `Group By` filter with `none` (default), `accounts` and `networks` grouping options.
- Implements loading state when loading balances. 
  - Loading state is only shown if hiding small balances, OR assets are grouped (groups with 0 total balance are hidden)

This pull request fixes #7765, fixes #7618

## Submitter Checklist:


## Test Plan:
1. Open Wallet Portfolio, it should not be grouped by default.
2. Tap Filters & Display Settings button and then tap `Reset`
3. Verify default `Group By` option is `None`.
4. Change `Group By` to accounts and tap `Save Changes`.
5. Verify Portfolio is now grouped by each account, with the assets below showing the balance only for that account.
    - Note: Groups with 0 total balance are hidden automatically if `Hide Small Balances` is enabled.
6. Verify groups are sorted by fiat value. 
7. Tap Filters & Display Settings button, then de-select some accounts and tap `Save Changes`
8. Verify de-selected account groups are no longer shown.
9. Tap Filters & Display Settings button then tap `Reset` (re-selects all accounts again), then change `Group By` to `Networks` and tap `Save Changes`
10. Verify Portfolio is now grouped by each network with only assets for that network shown in each group.
    - Note: Groups with 0 total balance are hidden automatically if `Hide Small Balances` is enabled.
11. Verify groups are sorted by fiat value. 
12. Tap Filters & Display Settings button then tap `Reset`.
13. De-select all accounts, or all networks and tap `Save Changes`.
14. Verify empty state is shown.
    - Empty state should be shown whenever no assets are displayed. This can occur with many combinations, for example hiding small balances, grouping assets and then hiding accounts with balances or hiding networks with balances, etc. 
15. Close the wallet, or quit Brave. Re-open Portfolio and verify `Group By` filter is preserved.


## Screenshots:

https://github.com/brave/brave-ios/assets/5314553/6d6e5536-53c2-4a78-8ce4-cf0e93a099b8


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
